### PR TITLE
pg sources: always return irrecoverable errors during snapshotting

### PR DESCRIPTION
**Pull request by sploiselle**
Going to cry wolf and say that I have a very compelling case for a solution to #18329 based on an
actual repro.

tl;dr -- indeterminate errors during snapshotting can lead to re-ingesting the snapshot, and we had
two spots where we were returning indeterminate (rather than irrecoverable) errors.

## Substance

A huge shout out to @def- for working what seemed like an incredibly long day to provide us a repro with [some logging enabled](https://github.com/sploiselle/materialize/commit/ceab6b1254ef28d9e85bdfa62d478c08fa8918c9). 🎉 

You can see the full repro logs attached.

Here is a highly edited version of the logs in question, where `u27` is the offending source:

```text
User(13): starting replication at replication_lsn 0/0
User(13): replication inner at replication_lsn 0/0
User(13): slot_lsn 0/1927430 snapshot_lsn 0/1927430
Health transition for source u13: None -> Running
User(13): snapshot produced 100 messages
replication snapshot for source u13 succeeded
User(13): starting replication slot at 0/1927430User(13): progressed to 0/1946819 with 100 messages
User(13): progressed to 0/19CED19 with 400 messages
...
replication for source u13 interrupted, retrying: connection closed
Health transition for source u13: Some(Running) -> StalledWithError { error: "connection closed", hint: None }
User(13): replication inner at replication_lsn 0/1A93A98
User(13): starting replication slot at 0/1A93A98
User(13): progressed to 0/1C76B99 with 0 messages
...
User(27): starting replication at replication_lsn 0/0
User(27): replication inner at replication_lsn 0/0
Health for source u27 being written to s659c9703-ee6e-4d4a-bb98-93ff3ecf3109
remap(u27) 3/4 initializing PersistHandle since=Antichain { elements: [0] } as_of=Antichain { elements: [0] } upper=Antichain { elements: [0] }
User(27): slot_lsn 0/25AA0B0 snapshot_lsn 0/25AA0B0
Health transition for source u27: None -> Running
User(27): snapshot produced 800 messages
replication for source u27 interrupted, retrying: connection closed
Health transition for source u27: Some(Running) -> StalledWithError { error: "connection closed", hint: None }
User(27): replication inner at replication_lsn 0/0
User(27): slot_lsn 0/25AA0B0 snapshot_lsn 0/25AA158
Health transition for source u27: Some(StalledWithError { error: "connection closed", hint: None }) -> Running
User(27): snapshot produced 800 messages
replication snapshot for source u27 succeeded
User(27): starting replication slot at 0/25AA0B0
...
```

There is a lot to unpack here.

Here are the few log messages that we really need to pay attention to:

- Dataflow starts:
  ```
  User(13): starting replication at replication_lsn 0/0
  ```
- Enters/reenters the replication loop:
  ```
  User(13): replication inner at replication_lsn 0/0
  ```

  Note that in the case where the LSN is `0/0`, we snapshot the collection.
- Encounters an indeterminate error:
  ```
  replication for source u13 interrupted, retrying: connection closed
  Health transition for source u13: Some(Running) -> StalledWithError { error: "connection closed", hint: None }
  ```
- Encounters an irrecoverable error:
  ```
  irrecoverable error for source u13...
  ```
- Snapshotting ends:
  ```
  replication snapshot for source u27 succeeded
  ```

The difference between indeterminate errors and irrecoverable errors is that the former will wait
and re-enter the replication loop, while the latter die forever and have to be torn
down/rescheduled. Irrecoverable errors are appropriate when you have potentially uncommitted data in
an in-memory buffer that you cannot simply drop, and is the appropriate type of error whenever
you're snapshotting; you might have sent a row to the `row_sender`, and you cannot "take it back."

So when snapshotting `u27`, you can see we begin snapshotting (implicitly), but then hit an
indeterminate error, and then snapshot again, and then succeed--but without encountering an
irrecoverable error inbetween, which means that the first snapshot is still buffered in memory.

This can be seen with the following log lines:

```
User(27): starting replication at replication_lsn 0/0
User(27): replication inner at replication_lsn 0/0
User(27): snapshot produced 800 messages
replication for source u27 interrupted, retrying: connection closed
User(27): snapshot produced 800 messages
replication snapshot for source u27 succeeded
```

For this to have not produced duplicate data, we would have needed `retrying` to be `irrecoverable`.

This theory is further substantiated by the fact that I found two spots in the code that returned
indeterminate rather than irrecoverable errors during snapshotting.

To fix this, I suggest we use a glorified try-catch that converts all indeterminate errors to
irrecoverable errors during snapshotting lest they sneak in again.

One complexity here is that I have no idea why the `pg_snapshot_failure` fail point did not catch
this regression. Maybe it isn't firing? The test that uses it waits and brings storage up, but I
don't know if it aggressively checks enough that the failpoint actually triggers.

### Motivation

This PR fixes a recognized bug. #18329

### Tips for reviewer

This is almost entirely code movement besides a few changes to also explicitly return irrecoverable
errors.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by guswynn** [comment]
> One complexity here is that I have no idea why the pg_snapshot_failure fail point did not catch
this regression. Maybe it isn't firing? The test that uses it waits and brings storage up, but I
don't know if it aggressively checks enough that the failpoint actually triggers.

It might be a timing issue on when the postgres replication stream is actually cut off? not sure tho

File: src/storage/src/source/postgres.rs [hunk 1]
```diff
@@ -525,184 +525,207 @@ async fn postgres_replication_loop(mut task_info: PostgresTaskInfo) {
...
+    let res = client
+        .simple_query(&format!(
+            r#"SELECT confirmed_flush_lsn FROM pg_replication_slots WHERE slot_name = '{}'"#,
+            task_info.slot
+        ))
+        .await?;
+    let slot_lsn = parse_single_row(&res, "confirmed_flush_lsn");
+    client
+        .simple_query("BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ;")
+        .await?;
```
Comment (guswynn) [inline src/storage/src/source/postgres.rs:567]: do we know what error was being erroneously marked as indefinite instead of irrecoverable? Should we get rid of this: https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/postgres.rs#L147-L155 impl and force every error in this module to explicitly use `map_err`?

Comment (sploiselle) [reply src/storage/src/source/postgres.rs:567]: That was my original approach but that still let people mess up, where as the try-catch is much harder to misuse, and with that in place it matters less. idk; I'm equivocal.

**Comment by sploiselle** [comment]
regarding

> One complexity here is that I have no idea why the pg_snapshot_failure fail point did not catch
> this regression. Maybe it isn't firing? The test that uses it waits and brings storage up, but I
> don't know if it aggressively checks enough that the failpoint actually triggers.

this diff shows the problem:

```diff
diff --git a/src/storage/src/source/postgres.rs b/src/storage/src/source/postgres.rs
index fedf12dd81..47a68af378 100644
--- a/src/storage/src/source/postgres.rs
+++ b/src/storage/src/source/postgres.rs
@@ -239,6 +239,7 @@ struct PostgresTaskInfo {
     row_sender: RowSender,
     sender: Sender<InternalMessage>,
     resume_lsn: Arc<AtomicU64>,
+    fail_once: bool,
 }
 
 impl SourceRender for PostgresSourceConnection {
@@ -336,6 +337,7 @@ impl SourceRender for PostgresSourceConnection {
                 row_sender: RowSender::new(dataflow_tx.clone()),
                 sender: dataflow_tx,
                 resume_lsn: Arc::clone(&resume_lsn),
+                fail_once: true,
             };
 
             task::spawn(|| format!("postgres_source:{}", config.id), {
@@ -601,15 +603,6 @@ async fn postgres_replication_loop_inner(
         );
 
         while let Some((i, event)) = stream.as_mut().next().await {
-            if i > 0 {
-                // Failure scenario after we have produced at least one row, but before a
-                // successful `COMMIT`
-                fail::fail_point!("pg_snapshot_failure", |_| {
-                    Err(ReplicationError::Indefinite(anyhow::anyhow!(
-                        "recoverable errors should crash the process"
-                    )))
-                });
-            }
             let (output, row) = match event {
                 Ok(event) => event,
                 Err(err @ ReplicationError::Definite(_)) => return Err(err),
@@ -623,6 +616,17 @@ async fn postgres_replication_loop_inner(
                 .await;
         }
 
+        if task_info.fail_once {
+            task_info.fail_once = false;
+            // Failure scenario after we have produced at least one row, but before a
+            // successful `COMMIT`
+            fail::fail_point!("pg_snapshot_failure", |_| {
+                Err(ReplicationError::Indefinite(anyhow::anyhow!(
+                    "recoverable errors should crash the process"
+                )))
+            });
+        }
+
         if let Some(temp_slot) = temp_slot {
             let _ = client
                 .simple_query(&format!("DROP_REPLICATION_SLOT {temp_slot:?}"))
diff --git a/test/cluster/mzcompose.py b/test/cluster/mzcompose.py
index 6303900fe6..bdbf82e794 100644
--- a/test/cluster/mzcompose.py
+++ b/test/cluster/mzcompose.py
@@ -992,12 +992,8 @@ def workflow_pg_snapshot_resumption(c: Composition) -> None:
         print("Sleeping to ensure that clusterd crashes")
         time.sleep(10)
 
-        with c.override(
-            # turn off the failpoint
-            Clusterd(name="storage")
-        ):
-            c.up("storage")
-            c.run("testdrive", "pg-snapshot-resumption/04-verify-data.td")
+        c.run("testdrive", "pg-snapshot-resumption/04-verify-data.td")
+
```

With this diff applied, the test will fail with a double-ingested snapshot, i.e. substantiating this solution as the right one to fix the bug IMO. However, it does not fail with an error that the server was brought down. This test was not doing what it thought it was doing: nothing checked that the fail point had brought the server down. The test itself was bringing the server down with its invocation of `up`, and that was then allowing the snapshot to get re-ingested.

I will deprecate the test if there are any further pushes to CI for this PR, or in a subsequent PR.

**Comment by def-** [comment]
> One complexity here is that I have no idea why the pg_snapshot_failure fail point did not catch
this regression. Maybe it isn't firing? The test that uses it waits and brings storage up, but I
don't know if it aggressively checks enough that the failpoint actually triggers.

Is there an explicit test for this? We do have PR code coverage now, so could check this: https://github.com/MaterializeInc/materialize/pull/18370#issuecomment-1506415653

**Comment by sploiselle** [comment]
@def- The issue here turned out to not be any issue with `failpoint!` but was instead an issue with the assumption we were making about the state that resulted from the `failpoint!`'s execution; more details in https://github.com/MaterializeInc/materialize/pull/18769#issuecomment-1507741879

**Review by petrosagg — APPROVED** [review]
Finally, thanks for getting to the bottom of this!

**Review by def- — APPROVED** [review]
Thanks for the fix! I haven't seen the issue with this change yet, so seems good to go from QA side.
